### PR TITLE
feat(react): Enable ThirdPartyAuth in React Signup

### PR DESCRIPTION
**Pull request by vpomerleau**
## Because

* We want to match parity with backbone Signup and allow users to sign up with third party authentication when enrolled in the React Signup experiment

## This pull request

* Adds the ThirdPartyAuth component to the React Signup form
* Builds on the initial scaffolding for the React ThirdPartyAuthCallback, but does not enable this React route to complete sign in after third party auth. The final steps are still handled by backbone until FXA-8834

## Issue that this pull request solves

Closes: #FXA-8761

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [ ] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

React signup form with ThirdPartyAuth options:
image

## Other information (Optional)

With this PR, it is expected that users will be able to:
- Sign up and sign in with Third Party Auth options whether or not they are enrolled in the React Signup experiment
- Third Party Auth options remain unavailable for both backbone and react if the signup is for Sync

### CODE REVIEW

File: packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx [hunk 1]
```diff
@@ -3,40 +3,150 @@
...
+import { AUTH_PROVIDER } from 'fxa-auth-client/browser';
+import { useAccount } from '../../../models';
+import {
+  StoredAccountData,
+  persistAccount,
+  setCurrentAccount,
+} from '../../../lib/storage-utils';
+import { sessionToken } from '../../../lib/cache';
+
+// TODO this page to be completed/activated in FXA-8834
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx:23]: Just a drive-by comment: when we do this, I think we're going to want to rework some of this to use `useValidatedQueryParams`.

File: packages/fxa-settings/src/components/ThirdPartyAuth/index.tsx [hunk 2]
```diff
@@ -200,23 +163,29 @@ const ThirdPartySignInForm = ({
...
+}
+
 function getState() {
   // We stash originating location in the state oauth param
   // because we will need it to use it to log the user into FxA
-  const currentParams = new URLSearchParams(window.location.search);
-  currentParams.delete('deeplink');
-  currentParams.set('showReactApp', 'true');
+  const originParams = new URLSearchParams(window.location.search);
+  const modifiedParams = deleteParams(originParams, [
```
Comment (vbudhram) [inline packages/fxa-settings/src/components/ThirdPartyAuth/index.tsx:175]: It is probably worth a comment here why we are removing these values, which is to ensure that the user is taken to Backbone verson after they auth with third party.

File: packages/fxa-settings/src/models/Account.ts [hunk 3]
```diff
@@ -917,15 +918,40 @@ export class Account implements AccountData {
 
   async verifyAccountThirdParty(
     code: string,
-    provider: AUTH_PROVIDER
+    provider: AUTH_PROVIDER = AUTH_PROVIDER.GOOGLE,
+    service?: string,
+    metricsContext: MetricsContext = {}
```
Comment (vbudhram) [inline packages/fxa-settings/src/models/Account.ts:923]: 👍🏽

File: packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx [hunk 4]
```diff
@@ -3,40 +3,150 @@
...
     }
+  };
 
-    navigate('/', { replace: true });
+  // Persist account data to local storage to match parity with content-server
+  // this allows the recent account to be used for /signin
+  const storeAccountData = (accountData: StoredAccountData) => {
+    persistAccount(accountData);
+    setCurrentAccount(accountData.uid);
+    sessionToken(accountData.sessionToken);
```
Comment (vbudhram) [inline packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx:108]: Would we also need to update graphql?

```
    await this.apolloClient.cache.writeQuery({
      query: GET_LOCAL_SIGNED_IN_STATUS,
      data: { isSignedIn: true },
    });
```

File: packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx [hunk 5]
```diff
@@ -3,40 +3,150 @@
...
+    } else {
+      // TODO validate what should happen if we hit this page
+      // without the required auth params to verify the account
+      hardNavigateToContentServer('/');
+    }
   }
 
   useEffect(() => {
-    handleOauthResponse();
+    verifyAndContinue();
```
Comment (vbudhram) [inline packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx:149]: Hm, not sure if either function name completely describes what is going on. The `ThirdPartyAuthCallback` page should verify the oauth response from Apple/Google, and then complete the sign-in based on what integration is being used.

**Review by vbudhram — APPROVED** [review]
@vpomerleau Just a few nits that I think could be addressed in the follow up PR. I tested locally and works as described 👍🏽

File: packages/fxa-auth-server/lib/routes/linked-accounts.ts [hunk 6]
```diff
@@ -501,6 +502,7 @@ export const linkedAccountRoutes = (
               .string()
               .required()
               .description(DESCRIPTION.providerUid),
+            email: isA.string().required().description(DESCRIPTION.email),
```
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-auth-server/lib/routes/linked-accounts.ts:505]: I'm returning the email so that the email stored in local storage is the email used for third party auth (vs the email first entered in our email first form), as it might be different, especially if the user is using a private relay. Perhaps there's another way to do this?

File: packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx (hunk 1 above)
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx:23]: Added a TODO comment for FXA-8834 (and updated the description there too!)

File: packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx (hunk 4 above)
Comment (vpomerleau) [reply packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx:108]: I had added this in `verifyAccountThirdParty`, but I think it needs to be reshuffled. We're missing the actual sign in that will be completed in FXA-8834/FXA-6488 and that's where it should go. Added a note.

File: packages/fxa-settings/src/components/ThirdPartyAuth/index.tsx (hunk 2 above)
Comment (vpomerleau) [reply packages/fxa-settings/src/components/ThirdPartyAuth/index.tsx:175]: Currently, they are taken to backbone no matter what based on our router setup - IIUC, the state here is only stashed to reconstruct the redirect URL for RP. We don't read from this stashed state to decide between backbone/react on return from Google/Apple. Please correct me if I'm wrong!

File: packages/fxa-content-server/server/lib/routes/react-app/index.js [hunk 7]
```diff
@@ -105,7 +105,10 @@ const getReactRouteGroups = (showReactApp, reactRoute) => {
 
     postVerifyThirdPartyAuthRoutes: {
       featureFlagOn: showReactApp.postVerifyThirdPartyAuthRoutes,
-      routes: reactRoute.getRoutes(['post_verify/third_party_auth/callback']),
+      routes: reactRoute.getRoutes([
+        'post_verify/third_party_auth/callback',
```
[the remark this answers is not in the document]
Comment (vpomerleau) [reply packages/fxa-content-server/server/lib/routes/react-app/index.js:109]: Doesn't hurt to add them - the `postVerifyThirdPartyAuthRoutes` remain off for the time being, but this makes it a bit clearer which routes are included.

**Review by vbudhram — APPROVED** [review]
@vpomerleau Changes look good 👍🏽, I restarted the tests since it was a SubPlat failure.
